### DISCUSSION TITLE
[Bug] 启动时连接 mongodb 失败, 未退出服务

### DISCUSSION THREAD

**Discussion by qwang1113** [midwayjs/midway#4097] [Q&A] [answered]
### Describe the problem（描述问题）

根据 mongoose [Error Handling](https://mongoosejs.com/docs/connections.html#error-handling), mongoose 第一次连接失败并不会自动重连

midwayjs [这里](https://github.com/midwayjs/midway/blob/main/packages/mongoose/src/manager.ts#L52) 也没判断是否是第一次连接的情况

因此, 如果服务启动时, 连接没成功, 此时服务不主动退出, 因为没有 mongo 连接, 依赖的 API 持续报错, 希望这里加上重试或者第一次连接报错则退出服务的机制

### Midway Versions（Midway 版本）

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(3.15.8)
✓ @midwayjs/axios(3.15.8)
✓ @midwayjs/bootstrap(3.15.8)
✓ @midwayjs/bull(3.15.8)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/cache-manager(3.15.8)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.15.8)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(not installed)
✓ @midwayjs/cross-domain(3.15.11)
✓ @midwayjs/decorator(3.15.8)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✖ @midwayjs/i18n(current: 3.13.7, allow: ["3.15.8","3.15.10","3.15.11"])
✓ @midwayjs/info(not installed)
✓ @midwayjs/jwt(not installed)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.15.8)
✓ @midwayjs/mongoose(3.15.8)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(3.15.8)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(3.15.8)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(not installed)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.15.11)
✓ @midwayjs/version(3.15.11)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.15.11)
✓ @midwayjs/ws(not installed)
************************************************************************************************************************
>> Check complete, found  1  problem.
>> Use -u to show update list that can be upgraded to the latest version.
>> Use -m to show updates that can be upgraded to the most compatible version.
>> Use -u -w or -m -w to write to the package.json file and update the lock file if it exists.
>> Please check the result above.
************************************************************************************************************************

**Comment by czy88840616** [comment, marked as answer]
这应该不算 bug
- 1、启动不成功是需要业务开发上线前保证的，这个在线上会由健康检查来处理，所以未启动成功不会影响实际业务
- 2、启动错误是否退出，由错误是否能捕获决定，如果是不可逆转的错误（process 捕获级别），才会退出
- 3、组件是否成功的检查，这个会在 v4 考虑加一个超时和检查机制，明确哪个组件没有启动成功
